pynitrokey/cli/storage.py

```diff
@@ -40,14 +50,112 @@ def enable_update(password):
 
     local_print("Enabling firmware update mode")
-    nks = NitrokeyStorage()
-    nks.connect()
-    try:
-        if nks.enable_firmware_update(password) == 0:
-            local_print("setting firmware update mode - success!")
-    except DeviceNotFound:
-        local_print("No Nitrokey Storage device found")
+    nks = connect_nkstorage()
+    if nks.enable_firmware_update(password) == 0:
+        local_print("setting firmware update mode - success!")
+
+
+@click.command()
+@click.option(
+    "-p",
+    "--password",
+    prompt="User PIN required",
+    hide_input=True,
+    help="User password",
+)
+def open_encrypted(password):
+    """Unlock the encrypted storage"""
+    nks = connect_nkstorage()
+    ret = nks.unlock_encrypted_volume(password)
+    if not ret.ok:
+        if ret == RetCode.WRONG_PASSWORD:
+            raise CliException("Wrong user PIN", support_hint=False)
+        else:
+            raise CliException(
+                "Unexpected error unlocking the encrypted storage {}".format(str(ret))
+            )
+
+
+@click.command()
+def close_encrypted():
+    """Lock the encrypted storage"""
+    nks = connect_nkstorage()
+    ret = nks.lock_encrypted_volume()
+    if not ret.ok:
+        raise CliException("Error closing the encrypted volume: {}".format(str(ret)))
+
+
+@click.command()
+@click.option(
+    "-p",
+    "--password",
+    prompt="Please input the hidden storage passphrase",
+    hide_input=True,
+    help="Hidden storage passphrase",
+)
+def open_hidden(password):
+    """Unlock an hidden volume"""
+    nks = connect_nkstorage()
+    ret = nks.unlock_hidden_volume(password)
+    if not ret.ok:
+        if ret == RetCode.WRONG_PASSWORD:
+            raise CliException("Wrong hidden storage passphrase", support_hint=False)
+        else:
+            raise CliException(
+                "Unexpected error unlocking the hidden storage: {}".format(str(ret))
+            )
+
+
+@click.command()
+def close_hidden():
+    """Lock the hidden volumes"""
+    nks = connect_nkstorage()
+    ret = nks.lock_hidden_volume()
+    if not ret.ok:
+        raise CliException("Error closing the hidden volume: {}".format(str(ret)))
+
+
+@click.command()
+@click.argument(
+    "slot",
+    type=int,
+)
+@click.argument(
+    "begin",
+    type=int,
+)
+@click.argument("end", type=int)
+@click.option(
+    "-p",
+    "--password",
+    prompt="Please input the hidden storage passphrase",
+    hide_input=True,
+    help="Hidden storage passphrase",
+)
+def create_hidden(slot, begin, end, password):
+    """Create an hidden volume
+
+    SLOT is the slot used for the hidden volume (0-3)\n
+    START is where the volume begins expressed in percent of total available storage (0-99)\n
+    END is where the volume ends expressed in percent of total available storage (1-100)\n
+    """
+    if not slot in [0, 1, 2, 3]:
+        raise CliException("Error: Slot must be between 0 and 3", support_hint=False)
```

Comment:
One potential problem here would be, that Nitrokey App numbers HVs from 1, not 0, which could result in overwriting data in slot by mistake. Let's change that to range [1,4].
Please keep in mind to keep compatibility with Nitrokey App in respect to inputs.

Reply:
Good catch! I see how that could be confusing.